File: Swift/GenericStructures/List.swift
final class Node<T: Comparable>{
    lazy var data:T;
    var next:Node? = nil;
    var prev:Node? = nil;

    init(m_data:T) {data = m_data;}
}

class List<T: Comparable>{
    private var front:Node? = nil;
    private var back:Node? = nil;
    private var m_size:Int = 0;

    func add(data:T){
        var node:Node = Node(data);
        if isEmpty(){
            self.front = node;
            self.back = node;
        }
        else{
            self.back?.next = node;
            node.prev = self.back;
            self.back = node;
        }
        self.m_size += 1;
    }

    func add_back(data:T){
        var node:Node = Node(data);
        if isEmpty(){
            self.front = node;
            self.back = node;
        }
        else{
            self.back?.next = node;
            node.prev = self.back;
            self.back = node;
        }
        self.m_size += 1;
    }

    func add_front(data:T){
        var node:Node = Node(data);
        if isEmpty(){
            self.front = node;
            self.back = node;
        }
        else{
            node.next = self.front;
            self.front?.prev = node;
            self.front = node;
        }
        self.m_size += 1;
    }

    func remove_front(){ // how GC is managed in swift?
        if isEmpty(){
            return;
        }
        self.front = self.front?.next;
        self.m_size -= 1;
    }

    func remove_back(){
        if isEmpty(){
            return;
        }
        self.back = self.back?.prev;
        self.m_size -= 1;
    }

    func remove(data:T){
        if isEmpty(){
            return;
        }
        else{
            var ptr:Node? = self.front;
            var prev:Node? = self.ptr;

            while (ptr?.next != nil){
                if(ptr!.data == data){
                    prev?.next = ptr?.next;
                    ptr?.next.prev = prev;

                    self.m_size -= 1;
                    return; 
                }
                prev = ptr;
                ptr = ptr?.next;
            }
        }
    }

    func display(){
        var ptr:Node? = self.front;
        while(ptr?.next != nil){
            print(" \(ptr!.data)");
            ptr = ptr?.next;
        }
    }

    func isEmpty()->Bool{return self.front == nil;}

    func size()->Int{return self.m_size; }
}

File: Java/GenericStructures/List.java
package GenericStructures;

public class List<T>{
    class Node{
        T data;
        Node prev = null;
        Node next = null;

        Node(T data){ this.data = data;}
    }

    private Node front = null;
    private Node back = null;
    private int size;

    public List(){
        this.size = 0;
    }

    public void add(T data){
        Node node = new Node(data);
        if(is_empty()){
            this.front = node;
            this.back  = node;
        }
        else{
            this.back.next = node;
            node.prev = this.back;
            back = node;
        }
        this.size += 1;
    }

    public void add_front(T data){
        Node node = new Node(data); 
        if(is_empty()){
            this.front = node;
            this.back = node;
        }
        else{
            node.next = this.front;
            this.front.prev = node;
            this.front = node;
        }
        this.size += 1;
    }

    public void add_back(T data){
        Node node = new Node(data);
        if(is_empty()){
            this.front = node;
            this.back  = node;
        }
        else{
            this.back.next = node;
            node.prev = this.back;
            back = node;
        }
        this.size += 1;
    }

    public void remove(T data){
        Node ptr = this.front;
        try {
            boolean is_element = false;

            while(ptr.next != null){
                if(ptr.data == data){
                    ptr.prev = ptr.prev.prev;
                    ptr.next = ptr.next.next;
                    is_element = true;
                    return;
                }
                ptr = ptr.next;
            }

            if(is_element == false) throw new Exception();

        } catch (Exception e) {
            System.out.println(String.valueOf(data) + " is not element of the list.");
            return;
        }
    }

    public void display(){
        Node ptr = this.front;
        
        while(ptr.next != null){
            System.out.print(String.valueOf(ptr.data) + " ");
            ptr = ptr.next;
        }
    }

    public T top(){return this.front.data; }
    
    public T bottom(){return this.back.data; }

    public boolean is_empty(){
        if(this.front ==  null) return true;
        return false;
    }

    public int size(){return this.size; }

}

class MainDoubly{
    public void main(String [] args){

    }
}